File: sdks/javascript/failure/cache.failure.ts
import type { CacheEntry } from "../entry";

export type CacheFailureStage =
  | "MEMORY_READ"
  | "MEMORY_WRITE"
  | "IDB_READ"
  | "IDB_WRITE"
  | "VALIDATION"
  | "SERIALIZATION";

export type CacheFailureAction = "BYPASS_TO_API" | "RETURN_STALE_AND_REFRESH" | "HARD_FAIL";

export interface CacheFailureDecision<T = unknown> {
  action: CacheFailureAction;
  staleEntry?: CacheEntry<T>;
  reason: string;
}

export interface CacheFailureInput<T = unknown> {
  stage: CacheFailureStage;
  error: unknown;
  staleEntry?: CacheEntry<T>;
  allowStale: boolean;
  failClosed?: boolean;
}

export interface FallbackExecutionResult<T = unknown> {
  source: "API" | "CACHE";
  value: T;
  recovered: boolean;
  reason: string;
}

export function toFailureReason(error: unknown): string {
  if (error instanceof Error) {
    return error.message;
  }
  if (typeof error === "string") {
    return error;
  }
  if (error && typeof error === "object") {
    try {
      return JSON.stringify(error);
    } catch {
      return "Unknown cache failure";
    }
  }
  return "Unknown cache failure";
}

/**
 * Deterministic failure classifier for cache fallback policy.
 */
export function classifyCacheFailure<T = unknown>(
  input: CacheFailureInput<T>
): CacheFailureDecision<T> {
  const reason = `${input.stage}: ${toFailureReason(input.error)}`;

  if (input.failClosed) {
    return {
      action: "HARD_FAIL",
      reason
    };
  }

  if (input.allowStale && input.staleEntry) {
    return {
      action: "RETURN_STALE_AND_REFRESH",
      staleEntry: input.staleEntry,
      reason
    };
  }

  return {
    action: "BYPASS_TO_API",
    reason
  };
}

/**
 * Executes API fallback based on deterministic failure decision.
 */
export async function executeWithApiFallback<T>(params: {
  decision: CacheFailureDecision<T>;
  fetchFromApi: () => Promise<T>;
}): Promise<FallbackExecutionResult<T>> {
  if (params.decision.action === "HARD_FAIL") {
    throw new Error(params.decision.reason);
  }

  if (params.decision.action === "RETURN_STALE_AND_REFRESH") {
    if (!params.decision.staleEntry) {
      throw new Error("Missing stale entry for RETURN_STALE_AND_REFRESH action");
    }

    return {
      source: "CACHE",
      value: params.decision.staleEntry.data,
      recovered: true,
      reason: params.decision.reason
    };
  }

  const value = await params.fetchFromApi();
  return {
    source: "API",
    value,
    recovered: true,
    reason: params.decision.reason
  };
}

File: sdks/flutter/lib/src/failure/cache_failure.dart
import '../entry/cache_entry.dart';

enum CacheFailureStage {
  memoryRead,
  memoryWrite,
  persistentRead,
  persistentWrite,
  validation,
  serialization,
}

enum CacheFailureAction {
  bypassToApi,
  returnStaleAndRefresh,
  hardFail,
}

class CacheFailureDecision<T> {
  const CacheFailureDecision({
    required this.action,
    required this.reason,
    this.staleEntry,
  });

  final CacheFailureAction action;
  final CacheEntry<T>? staleEntry;
  final String reason;
}

class CacheFailureInput<T> {
  const CacheFailureInput({
    required this.stage,
    required this.error,
    required this.allowStale,
    this.staleEntry,
    this.failClosed = false,
  });

  final CacheFailureStage stage;
  final Object error;
  final CacheEntry<T>? staleEntry;
  final bool allowStale;
  final bool failClosed;
}

enum FallbackSource { api, cache }

class FallbackExecutionResult<T> {
  const FallbackExecutionResult({
    required this.source,
    required this.value,
    required this.recovered,
    required this.reason,
  });

  final FallbackSource source;
  final T value;
  final bool recovered;
  final String reason;
}

String toFailureReason(Object error) {
  if (error is Error) {
    return error.toString();
  }
  if (error is Exception) {
    return error.toString();
  }
  return error.toString();
}

CacheFailureDecision<T> classifyCacheFailure<T>(CacheFailureInput<T> input) {
  final reason = '${input.stage.name}: ${toFailureReason(input.error)}';

  if (input.failClosed) {
    return CacheFailureDecision<T>(
      action: CacheFailureAction.hardFail,
      reason: reason,
    );
  }

  if (input.allowStale && input.staleEntry != null) {
    return CacheFailureDecision<T>(
      action: CacheFailureAction.returnStaleAndRefresh,
      staleEntry: input.staleEntry,
      reason: reason,
    );
  }

  return CacheFailureDecision<T>(
    action: CacheFailureAction.bypassToApi,
    reason: reason,
  );
}

Future<FallbackExecutionResult<T>> executeWithApiFallback<T>({
  required CacheFailureDecision<T> decision,
  required Future<T> Function() fetchFromApi,
}) async {
  switch (decision.action) {
    case CacheFailureAction.hardFail:
      throw StateError(decision.reason);
    case CacheFailureAction.returnStaleAndRefresh:
      final staleEntry = decision.staleEntry;
      if (staleEntry == null) {
        throw StateError(
            'Missing stale entry for returnStaleAndRefresh action');
      }
      return FallbackExecutionResult<T>(
        source: FallbackSource.cache,
        value: staleEntry.data,
        recovered: true,
        reason: decision.reason,
      );
    case CacheFailureAction.bypassToApi:
      final value = await fetchFromApi();
      return FallbackExecutionResult<T>(
        source: FallbackSource.api,
        value: value,
        recovered: true,
        reason: decision.reason,
      );
  }
}
